File: leetcode528/leetcode528_my_version.cpp
class Solution {
public:
    Solution(vector<int>& w) {
        aggregated_sum = vector<int> (w.size());
        
        for (int i = 0; i < w.size(); ++i) {
            sum += w[i];
            aggregated_sum[i] = sum;
        }
    }
    
    int pickIndex() {        
        int random = rand() % sum + 1;
        int ans = binarySearch(random, 0, aggregated_sum.size() - 1);
        return ans;
    }
private:
    int sum = 0;
    vector<int> aggregated_sum;
    
    int binarySearch(int random, int left, int right) {
        
        
        while (left < right) {
            int mid = left + (right - left) / 2;
            // get answer
            if (random == aggregated_sum[mid]) {
                return mid;
            }
            
            // left side
            else if (random < aggregated_sum[mid]) {
                right = mid;
            }
            
            // right side
            else if (random > aggregated_sum[mid]) {
                left = mid + 1;
            }
        }
        
        return left;
    }
};

/**
 * Your Solution object will be instantiated and called as such:
 * Solution* obj = new Solution(w);
 * int param_1 = obj->pickIndex();
 */

// Runtime: 332 ms, faster than 21.28% of C++ online submissions for Random Pick with Weight.
// Memory Usage: 40.1 MB, less than 95.41% of C++ online submissions for Random Pick with Weight.
// time complexity: Solution O(n) pickIndex O(log n)
// space copmlexity: O(n)

File: leetcode528/leetcode528_my_version.java
class Solution {
    List<Integer> arrL;
    
    public Solution(int[] w) {
        arrL = new ArrayList<>();
        int accumulatedSum = 0;
        arrL.add(0);
        for (int i = 0; i < w.length; i++)
        {
            accumulatedSum += w[i];
            arrL.add(accumulatedSum);
        }
    }
    
    
    // w = [1, 2, 3, 4, 5, 6]
    // arrL = {0, 1, 3, 6, 10, 15, 21}
    public int pickIndex() {
        // List<Integer> tmp = new ArrayList<>(arrL);
        
        Random r = new Random();
        int rand = r.nextInt(arrL.get(arrL.size() - 1));
        
        // binary search
        int left = 0, right = arrL.size();
        int ans = 0;
        
        while (left < right)
        {
            int mid = left + (right - left) / 2;
            
            if (arrL.get(mid) == rand)
            {
                return mid;
            }
            else if (arrL.get(mid) < rand)
            {
                ans = mid;
                left = mid + 1;
            }
            else
            {
                right = mid;
            }
        }
        
        return ans;
    }
}

/**
 * Your Solution object will be instantiated and called as such:
 * Solution obj = new Solution(w);
 * int param_1 = obj.pickIndex();
 */

// Runtime: 27 ms, faster than 31.43% of Java online submissions for Random Pick with Weight.
// Memory Usage: 43.5 MB, less than 92.71% of Java online submissions for Random Pick with Weight.
// time complexity: O(log n)
// space complexity: O(n)